[DM-15396] Stack based container builds

cbanek opened the pull request:
Here's a sneak preview of the stack based container builds.

A few changes:
1) Using the weekly stack container that sqre builds.  One small note, this is centos7 instead of ubuntu, so a few things changed.
2) I'm running instructions as root, but telling uwsgi to run as the lsst user, which already exists now in the container as the default stack user.  No need to create a user and chmod things as before.  No need to su run uwsgi.
3) Installs all the python modules using setup.py
4) Change nginx start procedure.  Centos containers don't even have systemctl installed by default, but running nginx will spin up the server and its workers.  It is now also configured to run as lsst.
5) Don't bother redirecting stdout to a file.  Maybe I should tee it, but having the output right there in kubectl logs -f is super useful.

Things left:
1) Imgserv unit tests don't work yet, due to dependency on lsst.db eups package.
2) Don't call out a specific weekly stack image in the dockerfile, but track the latest successful one.

cbanek committed:
[DM-15396] Add the start of a new stack based container.

This will be the host container for web services going forward
that use the stack for things (like imgserv) and then become
a web service.

cbanek committed:
[DM-15396] Stack container: checkpoint.

Got everything mostly stitched together.  Things run as the lsst
user, and we're using the stack container.  Install python modules
on their branches where I've been working from.

One last problem is getting the imgserv tests to run as part of
creating the container, and then after the merge the branches
can be changed to master.

brianv0 commented on docker/stackbase/webserv/entrypoint.sh:
```diff
@@ -0,0 +1,12 @@
+source /opt/lsst/software/stack/loadLSST.bash
+setup lsst_distrib
+uwsgi --master --processes 40 --threads 1 \
+    --socket /tmp/webserv.sock --uid lsst \
```
We should go ahead and keep it  this way for now, but just a note for the future:

We should be able to operate uwsgi in HTTP mode and get rid of nginx in this container completely:
https://uwsgi-docs.readthedocs.io/en/latest/HTTP.html

cbanek replied:
Yeah I was wondering about that as well, but that can be a later commit for sure.  I don't like nginx there, it's just another layer for things to get lost and have to configure!

brianv0 commented on docker/stackbase/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+
+# Install packages and run additional commands.
+COPY bake.sh /root/bake.sh
+RUN /root/bake.sh
```
I was thinking something a bit more along the lines of this of treating `dax_webserv` also as a python package with a requirements.txt file, so you could ideally do something along these lines:

```Dockerfile
FROM lsstsqre/centos:7-stack-lsst_distrib-w_2018_33
LABEL maintainer="cbanek@lsst.org"

# Not sure if root is necessary 
USER root 

RUN yum update -y && \
    yum install -y  --setopt=tsflags='' mysql-devel nginx && \
    yum clean all

RUN chown -R lsst:lsst /webserv
RUN chown -R lsst:lsst /var/lib/nginx

# Configure nginx
COPY etc/nginx/nginx.conf /etc/nginx/nginx.conf
COPY etc/nginx/sites-enabled/webserv /etc/nginx/sites-enabled/webserv

COPY --chmod=lsst:lsst bin python requirements.txt setup.py /webserv

# Note - installation is not in this directory but ends up in miniconda
# so we might want to do this as `USER lsst`... not sure.
RUN source /opt/lsst/software/stack/loadLSST.bash && \
    pip install -r requirements.txt

# Bring in ENTRYPOINT
COPY webserv/entrypoint.sh /webserv/entrypoint.sh

# Switch to root so we can start nginx,
# entrypoint does a `su webserv` internally to start wsgi as
# as webserv

ENTRYPOINT /webserv/entrypoint.sh
# Install packages and run additional commands.
```

... where requirements.txt could actually refer to the repositories for the `dax_` dependencies, or the specific commits

(from https://pip.readthedocs.io/en/1.1/requirements.html#git)

I think the requirements.txt file might look like
```
uwsgi
git+https://github.com/lsst/dax_imgserv.git@master#egg=dax_imgserv
git+https://github.com/lsst/dax_metaserv.git@master#egg=dax_metaserv
git+https://github.com/lsst/dax_dbserv.git@master#egg=dax_dbserv
git+https://github.com/lsst/dax_webservcommon.git@master#egg=dax_webservcommon
```

brianv0 replied:
This would also mean we'd have the `Dockerfile` at the root of the repo in this scenario

cbanek replied:
I get what you're talking about, but there are some subtle points to the reasons why I've done it the way I've done it.  In the end, I feel like it's mostly the same, in that the python packages are installed, but the way I've done it I consider it more developer friendly.

1. I really like having the source code around, like in the git repository.  It allows you to go there, go back and forth, and do a simple git diff to see if anyone has been editing the files or messed with them after the fact.  Super useful.  If I do the requirements.txt route, it doesn't actually keep the repos anywhere (it puts them in /tmp, then cleans them up by the time they are finally installed).  This is also super useful as it is the developer ready environment, where there are git repos and you can make changes and reinstall to test things out on your dev environment and have everything right there.

2. Using a requirements.txt won't actually run the tests, which is also pretty important.  I like the idea of running the tests as a part of blessing a container for production, so I think that should stay.

3. In the (hopefully very near) future, this whole problem should be greatly simplified by getting rid of stuff we don't want, and stuff we don't need.  I have tried my best to keep things status quo in this change, just to get the CI system primed to take bigger changes.  But what I hope to do in the future is:

3a. Get rid of db serv v0, and move that repo into some junkyard or something.  It seems like albuquery has replaced it, but we're still having to carry around this db v0 for no reason that I can see.
3b. We want to move metaserv out of here, and into some kind of TAP_SCHEMA protocol, which will likely look more like albuquery than its current form.  That also gets rid of this server.
3c. Now we're just left with imgserv, and webserv, and webservcommon (and also webservclient, although if we're using VO standards, can we get rid of this too?).  Imgserv has the actual server, webserv is some scaffolding to run mutliple services together (which we don't need), and common is common things among the services, and now there's only one.  We should be able to compress this all down eventually to one repo.

4. I understand the desire to pull things right out of the webserv repo rather than reclone in the docker build process, because it allows you to test uncommitted changes.  Sadly for any other repo (which is more likely where the logic will be) you still have to push a branch to test it out.  If we were down to one repo, using this methodology would be ideal, since we'd just pull everything from the local checkout and have no more dependency nightmares.

brianv0 replied:
Yes that's fine I think. We can work on merging the repo in the long run. 

```
- tox
- flake8
- etc...
- docker build -t ... .
```

We don't have to use the requirements.txt - [we can put it in the setup.py script](https://python-packaging.readthedocs.io/en/latest/dependencies.html#packages-not-on-pypi), I'm not that worried about it, and that still can get us a requirements.txt at deployment time if we want.

W.R.T. running the tests, we will be doing this in travis/jenkins in which case it'd be nice to install time but at a different time so we can abort before we even try to do anything with building the container. But, there's a potential bootstrapping problem in that you want a container/environment to run your tests in (and it's possibly convenient or necessary if that environment is very close to the container you are going to deploy)

I think if this is working we should just merge now and focus on consolidating repos and rethink at that point.

cbanek committed:
[DM-15396] Update branches to master, note failing test.

There's some test failures that are blocking us, but for now just
don't run the tests until we fix the associated ticket.

cbanek commented:
Okay, rebased and made a commit to change all the branch names off of their ticket branches to master.  Noted the ticket number for the failing test.  This is ready to go in if people are okay with that.

cbanek committed:
[DM-15396] Note todo about base image.

There's another ticket DM-15620 to track getting the latest weekly
stack tag and putting it in here.
